File: frontend/lib/models/order.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';

part 'order.g.dart';

@JsonSerializable()
class Order extends Equatable {
  final String id;
  final String userId;
  final String loungeId;
  final List<OrderItem> items;
  final double totalPrice;
  final String status;
  final String paymentMethod;
  final String? paymentId;
  final String? contractId;
  final String qrCode;
  final String? qrCodeImage;
  final DateTime? estimatedReadyTime;
  final DateTime? deliveredAt;
  final double commission;
  final String? notes;
  final String? cancellationReason;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Order({
    required this.id,
    required this.userId,
    required this.loungeId,
    required this.items,
    required this.totalPrice,
    required this.status,
    required this.paymentMethod,
    this.paymentId,
    this.contractId,
    required this.qrCode,
    this.qrCodeImage,
    this.estimatedReadyTime,
    this.deliveredAt,
    required this.commission,
    this.notes,
    this.cancellationReason,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) => _$OrderFromJson(json);
  Map<String, dynamic> toJson() => _$OrderToJson(this);

  @override
  List<Object?> get props => [
        id,
        userId,
        loungeId,
        items,
        totalPrice,
        status,
        paymentMethod,
        paymentId,
        contractId,
        qrCode,
        qrCodeImage,
        estimatedReadyTime,
        deliveredAt,
        commission,
        notes,
        cancellationReason,
        createdAt,
        updatedAt,
      ];
}

@JsonSerializable()
class OrderItem extends Equatable {
  final String foodId;
  final String name;
  final int quantity;
  final double price;
  final double subtotal;
  final int? estimatedTime;

  const OrderItem({
    required this.foodId,
    required this.name,
    required this.quantity,
    required this.price,
    required this.subtotal,
    this.estimatedTime,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) =>
      _$OrderItemFromJson(json);
  Map<String, dynamic> toJson() => _$OrderItemToJson(this);

  @override
  List<Object?> get props => [foodId, name, quantity, price, subtotal, estimatedTime];
}

// Order Status
enum OrderStatus {
  pending,
  preparing,
  ready,
  delivered,
  cancelled,
}

extension OrderStatusExtension on OrderStatus {
  String get displayName {
    switch (this) {
      case OrderStatus.pending:
        return 'Pending';
      case OrderStatus.preparing:
        return 'Preparing';
      case OrderStatus.ready:
        return 'Ready for Pickup';
      case OrderStatus.delivered:
        return 'Delivered';
      case OrderStatus.cancelled:
        return 'Cancelled';
    }
  }
}

File: backend/src/models/Order.js
const mongoose = require('mongoose');

const orderSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'User is required']
  },
  loungeId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Lounge',
    required: [true, 'Lounge is required']
  },
  items: [{
    foodId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Food',
      required: true
    },
    name: {
      type: String,
      required: true
    },
    quantity: {
      type: Number,
      required: true,
      min: 1
    },
    price: {
      type: Number,
      required: true,
      min: 0
    },
    subtotal: {
      type: Number,
      required: true,
      min: 0
    },
    estimatedTime: {
      type: Number,
      comment: 'Estimated preparation time in minutes'
    }
  }],
  totalPrice: {
    type: Number,
    required: [true, 'Total price is required'],
    min: 0
  },
  status: {
    type: String,
    enum: ['pending', 'preparing', 'ready', 'delivered', 'cancelled'],
    default: 'pending'
  },
  paymentMethod: {
    type: String,
    enum: ['contract', 'chapa'],
    required: [true, 'Payment method is required']
  },
  paymentId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Payment'
  },
  contractId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Contract'
  },
  qrCode: {
    type: String,
    unique: true,
    sparse: true
  },
  qrCodeImage: {
    type: String
  },
  estimatedReadyTime: {
    type: Date
  },
  deliveredAt: {
    type: Date
  },
  commission: {
    type: Number,
    default: 0,
    min: 0
  },
  notes: {
    type: String
  },
  cancellationReason: {
    type: String
  }
}, {
  timestamps: true
});

// Index for efficient queries
orderSchema.index({ userId: 1, status: 1 });
orderSchema.index({ loungeId: 1, status: 1 });
orderSchema.index({ qrCode: 1 });
orderSchema.index({ createdAt: -1 });

// Calculate estimated ready time before saving
orderSchema.pre('save', function(next) {
  if (this.isNew && !this.estimatedReadyTime) {
    // Calculate based on items' estimated times
    const maxEstimatedTime = Math.max(...this.items.map(item => item.estimatedTime || 15));
    this.estimatedReadyTime = new Date(Date.now() + maxEstimatedTime * 60000);
  }
  next();
});

module.exports = mongoose.model('Order', orderSchema);
